### DISCUSSION TITLE
Access CA without the use of Tokens

### DISCUSSION THREAD

**Discussion by prosticoco** [smallstep/certificates#533] [Q&A]
Hello, I would like to setup a PKI for an organization which would include multiple CAs (for fine-grained access control reasons) which would be responsible for issuing client-side certificates. Ideally, a client would register by authenticating with user/pwd to a preexisting web server which should then contact the corresponding CA (depending on the user) to issue a certificate.

 To this end, I would want to run multiple instances of `step-ca` (all on the same machine as the web server) and be able to access their functionalities (sign,revoke etc..) programmatically from an _https_ _go_ server using the _go_ client wrapper. 

Ideally, for readibility/cleanliness reasons, I want to avoid using/hardcoding bash commands to set the STEPPATH/ generate Tokens) and would just want to access the CAs using mTLS (for authenticating the client) in _go_.  However, I did not find any documentation for accessing the CAs functionality via https without manually generating tokens using the CLI.

Thank you for your help!

**Comment by prosticoco** [comment]
Alternatively, I can use the tokens but I need a way to generate them programmatically

**Reply by mmalone** [reply]
The tokens are actually just JWTs (RFC 7519) so you can absolutely generate them programmatically. I'd recommend either depending on our CLI code or using the https://github.com/square/go-jose package that we use to generate JWTs in golang. JWTs are notoriously easy to misuse though. So be careful.

It may be informative to generate a token and pipe it through our `step crypto jwt inspect` command to get a better idea of what we require in the JWT payload:

```
$ step ca token mike@smallstep.com --provisioner mike@smallstep.com | step crypto jwt inspect --insecure
✔ Provisioner: mike@smallstep.com (JWK) [kid: -6v0dwnSlz2D5opq-B5prQmyD9DhzdJSHMidiu1TjYg]
✔ Please enter the password to decrypt the provisioner key:
{
  "header": {
    "alg": "ES256",
    "kid": "-6v0dwnSlz2D5opq-B5prQmyD9DhzdJSHMidiu1TjYg",
    "typ": "JWT"
  },
  "payload": {
    "aud": "https://127.0.0.1:443/1.0/sign",
    "exp": 1618322940,
    "iat": 1618322640,
    "iss": "mike@smallstep.com",
    "jti": "3e15cb0ff77c49f047f07d436e71db0c16790167c6fa90878c00ed1ed9309d85",
    "nbf": 1618322640,
    "sans": [
      "mike@smallstep.com"
    ],
    "sha": "fab0330563738e29d368631eb2d548215c64be75511be3c50f5392d9ee4e68d9",
    "sub": "mike@smallstep.com"
  },
  "signature": "nRduTqjQwWpfsHMPXgLUXXn_iGstESRbHBkIc8N9TsOh7uVYbK-MmSP7QbD4muoXS0yAqGtqv0Eh7pKSh59PKg"
}
```

Note the `ES256` algorithm type (`alg`). The `aud` (audience) must be the full-path to the CA's sign URL. We check `exp` (expiration) and `nbf` (not before) times, so set those correctly. Tokens are single-use and we store the `jti` (JWT ID) to enforce that constraint. The `iss` (issuer) must match the name of the JWK provisioner that signed the JWT. The `sub` must match the common name in your certificate signing request, and `sans` must match your CSR subject alternative names.

Check out the CLI source at https://github.com/smallstep/cli/blob/master/command/ca/token.go for all the gory implementation details. You probably only need 10% of that code though if you're just doing JWT-based single-use token provisioning for one provisioner.

Based on the description of your problem, I am curious whether you could use the OIDC provisioner instead. It uses "identity tokens" generated via the OAuth OIDC single sign-on protocol and is designed for the use case you're describing (to delegate authentication to an identity provider using a web-based authentication flow for humans).

**Reply by prosticoco** [reply]
@mmalone Thanks for your detailed answer! _JWTs are notoriously easy to misuse though. So be careful._ exactly for this reason that I would rather trust your code :) I have actually found an easier way to generate them by directly creating a provisioner:
```go
        client,err := ca.NewClient("https://localhost:8443",ca.WithRootSHA256(fingerPrint))
	if err != nil {
		panic(err)
	}
	
	provisioner,err := ca.NewProvisioner("adm@ca.com","","https://localhost:8443",[]byte("1234"),ca.WithRootSHA256(fingerPrint))
	if err != nil {
		panic(err)
	}
	token,err := provisioner.Token("test_subject")
	if err != nil {
		panic(err)
	}
	req,private,err := ca.CreateSignRequest(token)
	if err != nil {
		panic(err)
	}
```
This enables me to generate valid certificates without using the CLI (except for initial setup). Thanks for letting me know if you see a big issue in this method.

**Reply by prosticoco** [reply]
Maybe I did not understand or still lack knowledge about OAuth/OIDC. From the docs : _In other words, you can use single sign-on with Google Workspace, Okta, Azure Active Directory, or any other OAuth OIDC provider to get a certificate._ : The organization I am working for use basic authentication on an https web server. The easiest way to me to connect this web server to a "backend" group of multiple CAs is by having another internal golang server which takes care of managing the different CAs. This CA server is then automatically accessed locally by the web server when administrator authenticate successfully to generate certificates.

**Reply by mmalone** [reply]
Yea, depending on our code as you've done is probably the easiest answer here.

The OAuth OIDC protocol is a "delegated authentication" mechanism. It standardizes a workflow where an OAuth _client_ sends a _user_ to an _identity provider_. The user authenticates to the identity provider, and the identity provider issues a signed assertion (a JWT) that basically says "the identity of the user presenting this token is _name_".

Google, Okta, Azure AD, etc. all implement the OAuth OIDC protocol. But you can also run your own OAuth OIDC server (your own _identity provider_) and authenticate users however you choose (e.g., using HTTP Basic authentication). There are open source OIDC providers you can run like https://github.com/dexidp/dex, as well as commercial providers like Okta and Auth0.

You're right that the OAuth OIDC bit would not fix your multi-ca problems. You'd still need some shim to handle that (either a new client or a proxy, as you've proposed).

This may all be overkill for your use case. OAuth OIDC was really designed to allow one piece of software to delegate authentication to another piece of software such that the first piece of software never sees the user's password. If you don't have that problem then you probably don't need OAuth OIDC.

**Reply by prosticoco** [reply]
Okay, OAuth OIDC is still an option as well then. Many thanks for all the great explanations/pointers!
